File: octave/c002h.m
% Solution to /r/dailyprogrammer Challenge #2 Hard
% Jimmy Nguyen
%
% Prompt:
%     Create a stopwatch program. This program should have start, stop, and lap
% options, and it should write out to a file to be viewed later.
%
% Link to challenge:
%     https://www.reddit.com/r/dailyprogrammer/comments/pjsdx/difficult_challenge_2/
%
% Usage:
%     help c002h
%         - prints help menu
%
%     c002h
%         - launches program
function c002h()
	filename = '../trash/c002h.log';
	fprintf('\n==================================================');
	fprintf('\n> STOPWATCH (%s)', filename);
	fprintf('\n==================================================');
	days_to_seconds = 24*3600;
	is_terminate = false;
	prompt_user = false;
	start_time = [];
	command = '';
	while ~is_terminate
		if prompt_user
			command = input('> ', 's');
			fprintf('--------------------------------------------------');
		end
		if isempty(start_time)
			if strcmpi(command, 'q')
				is_terminate = true;
			elseif strcmp(command, 's')
				start_time = now();
				write_to_file(filename, sprintf('started: %s', date_str(start_time)));
				prompt_user = false;
			else
				prompt_user = validate_command(prompt_user, command);
				fprintf('\n> enter s to start');
				fprintf('\n> enter q to quit');
				fprintf('\n--------------------------------------------------');
				fprintf('\n');
			end
		else
			end_time = now();
			end_datestr = date_str(end_time);
			seconds_passed = (end_time - start_time)*days_to_seconds;
			if strcmpi(command, 'e')
				write_to_file(filename, sprintf('ended  : %s - %.4f seconds passed', end_datestr, seconds_passed));
				start_time = [];
				prompt_user = false;
			elseif strcmpi(command, 'l')
				write_to_file(filename, sprintf('lapped : %s - %.4f seconds passed', end_datestr, seconds_passed));
				start_time = end_time;
			else
				prompt_user = validate_command(prompt_user, command);
			end
			if ~isempty(start_time)
				fprintf('\n> enter e to stop');
				fprintf('\n> enter l to lap');
				fprintf('\n--------------------------------------------------');
				fprintf('\n');
			end
		end
	end
	fprintf('\n');
	fprintf('\n');

function date_num_str = date_str(date_num)
	date_num_str = datestr(date_num, 'mm/dd/yyyy HH:MM:SS');

function write_to_file(filename, message)
	fh = fopen(filename, 'a');
	fprintf(fh, '%s\n', message);
	fclose(fh);
	fprintf('\n%s', message);

function prompt_user = validate_command(prompt_user, command)
	if ~prompt_user
		prompt_user = true;
	else
		fprintf('\n> invalid command "%s". please enter a valid command:', command);
	end

File: c++/c002h.cpp
/*
 * Solution to /r/dailyprogrammer Challenge #2 Hard
 * Jimmy Nguyen
 *
 * Prompt:
 *     Create a stopwatch program. This program should have start, stop, and lap
 * options, and it should write out to a file to be viewed later.
 *
 * Link to challenge:
 *     https://www.reddit.com/r/dailyprogrammer/comments/pjsdx/difficult_challenge_2/
 *
 * Compilation:
 *     g++ c002h.cpp
 *
 * Usage:
 *     ./a.out c002h
 *         - launches program
 */
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <sys/time.h>

#define FILENAME "../trash/c002h.log"

std::string input(std::string prompt)
{
	std::string input;
	std::cout << prompt;
	std::getline(std::cin, input);
	return input;
}

std::string to_string(time_t time)
{
	char buffer [20];
	std::strftime(buffer, 20, "%m/%d/%Y %H:%M:%S", std::localtime(&time));
	std::string time_str(buffer);
	return time_str;
}

std::string to_string(double d)
{
	std::stringstream ss;
	ss << std::fixed << std::setprecision(4) << d;
	return ss.str();
}

void write_to_file(std::string message)
{
	std::ofstream file;
	file.open(FILENAME, std::ios_base::app);
	file << message << std::endl;
	std::cout << std::endl << message;
}

bool validate_command(bool prompt_user, std::string command)
{
	if (prompt_user)
		std::cout << std::endl
				<< "> invalid command \""
				<< command
				<< "\". please enter a valid command:";
	else
		prompt_user = true;
	return prompt_user;
}

int main(int argc, char *argv[])
{
	std::cout << std::endl << "==================================================";
	std::cout << std::endl << "> STOPWATCH (" << FILENAME << ")";
	std::cout << std::endl << "==================================================";
	bool is_terminate = false;
	bool prompt_user = false;
	struct timeval start_time_tv, end_time_tv;
	time_t start_time = -1, end_time;
	std::string command, end_datestr;
	double milliseconds_passed;
	while (!is_terminate)
	{
		if (prompt_user)
		{
			command = input("> ");
			std::cout << "--------------------------------------------------";
		}
		if (start_time == -1)
		{
			if (command.compare("q") == 0)
				is_terminate = true;
			else if (command.compare("s") == 0)
			{
				gettimeofday(&start_time_tv, NULL);
				start_time = start_time_tv.tv_sec;
				write_to_file("started: " + to_string(start_time));
				prompt_user = false;
			}
			else
			{
				prompt_user = validate_command(prompt_user, command);
				std::cout << std::endl << "> enter s to start";
				std::cout << std::endl << "> enter q to quit";
				std::cout << std::endl << "--------------------------------------------------";
				std::cout << std::endl;
			}
		}
		else
		{
			gettimeofday(&end_time_tv, NULL);
			end_time = end_time_tv.tv_sec;
			end_datestr = to_string(end_time);
			milliseconds_passed = (end_time_tv.tv_sec - start_time_tv.tv_sec)*1000 + (end_time_tv.tv_usec - start_time_tv.tv_usec)/1000;
			if (command.compare("e") == 0)
			{
				write_to_file("ended  : " + end_datestr + " - " + to_string(milliseconds_passed/1000) + " seconds passed");
				start_time = -1;
				prompt_user = false;
			}
			else if (command.compare("l") == 0)
			{
				write_to_file("lapped : " + end_datestr + " - " + to_string(milliseconds_passed/1000) + " seconds passed");
				start_time = end_time;
			}
			else
				prompt_user = validate_command(prompt_user, command);
			if (start_time != -1)
			{
				std::cout << std::endl << "> enter e to stop";
				std::cout << std::endl << "> enter l to lap";
				std::cout << std::endl << "--------------------------------------------------";
				std::cout << std::endl;
			}
		}
	}
	std::cout << std::endl << std::endl;
	return 0;
}
